docs/spec/annotations.rst

```diff
@@ -53,42 +53,151 @@ decorators ``@property``, ``@staticmethod`` and ``@classmethod``.
-by a special comment, or in a separately distributed stub
-file.  (See below for examples.)
+Type and annotation expressions
+-------------------------------
+
+The terms *type expression* and *annotation expression* denote specific
+subsets of Python expressions that are used in the type system.
+
+.. _`type-expression`:
+
+A *type expression* is any expression that validly expresses a type. Type
+expressions are always acceptable in annotations and also in various other
+places. Specifically, type expressions are used in the following locations:
+
+* In a type annotation (always as part of an annotation expression)
+* The first argument to :ref:`cast() <cast>`
+* The second argument to :ref:`assert_type() <assert-type>`
+* The bounds and constraints of a ``TypeVar`` (whether created through the
+  old syntax or the native syntax in Python 3.12)
+* The definition of a type alias (whether created through the ``type`` statement,
+  the old assignment syntax, or the ``TypeAliasType`` constructor)
+* The type arguments of a generic class (which may appear in a base class
+  or in a constructor call)
+* The definitions of fields in the functional forms for creating
+  :ref:`TypedDict <typeddict>` and :ref:`NamedTuple <namedtuple>` types
+* The base type in the definition of a :ref:`NewType <newtype>`
+
+.. _`annotation-expression`:
+
+An *annotation expression* is an expression that is acceptable to use in
+an annotation context (a function parameter annotation, function return
+annotation, or variable annotation). Generally, an annotation expression
+is a type expression, optionally surrounded by one or more :term:`type qualifiers <type qualifier>`
+or by `Annotated`. Each type qualifier is valid only in some contexts. Note
+that while annotation expressions are the only expressions valid as type
+annotations in the type system, the Python language itself makes no such
+restriction: any expression is allowed.
 
 Annotations must be valid expressions that evaluate without raising
-exceptions at the time the function is defined (but see below for
-forward references).
-
-Annotations should be kept simple or static analysis tools may not be
-able to interpret the values. For example, dynamically computed types
-are unlikely to be understood.  (This is an
-intentionally somewhat vague requirement; specific inclusions and
-exclusions may be added in the future as warranted by the discussion.)
-
-In addition to the above, the following special constructs defined
-below may be used: ``None``, ``Any``, ``Union``, ``Tuple``,
-``Callable``, all ABCs and stand-ins for concrete classes exported
-from ``typing`` (e.g. ``Sequence`` and ``Dict``), type variables, and
-type aliases.
+exceptions at the time the function is defined (but see :ref:`forward-references`).
+
+.. _`expression-grammar`:
+
+The following grammar describes the allowed elements of type and annotation expressions:
+
+.. productionlist:: expression-grammar
```

Comment:
The rendered version of this block requires a lot of horizontal scrolling. Maybe manually break up the longer lines?

Reply:
Broke up the lines so that it no longer scrolls horizontally on my laptop.